### ISSUE TITLE
Gradient of regularised OT cost w.r.t. positions of mass is very different to that computed analytically for squared Euclidian distance cost.

### ISSUE THREAD

**Issue by ersisimou** [google-research/ott#9]
Hi! Many thanks for this package and its updates.

In the case where the cost for the transportation of mass is the squared Euclidian distance, one can compute the gradient of the regularised OT cost w.r.t. the positions of mass using the expression (9.7) in https://arxiv.org/pdf/1803.00567.pdf

I noticed that the same gradient when computed with jax.grad (with implicit differentiation or backpropagation and log stabilized or not) returned very different values.

I share a [collab](https://colab.research.google.com/drive/1YR1B5wX-8u4ZsO5vLl9z4HU5rvmmfx7j?usp=sharing) demonstrating this. The code for the closed form computation of the gradients is taken from [here](https://nbviewer.jupyter.org/github/gpeyre/numerical-tours/blob/master/python/optimaltransp_6_entropic_adv.ipynb)
and the code for the computation with jax.grad on the sinkhorn output from [this](https://github.com/google-research/ott/blob/master/tests/core/sinkhorn_differentiability_test.py) test. It seems that the gradients returned by jax.grad have significantly smaller values.

Were you expecting this difference? 

Thank you in advance for the response.

**Comment by marcocuturi** [comment]
Thanks so much for doing these tests, I want to insist that this is really helpful for us :) 

So, it seems the difference comes from a difference scaling. This is mentioned in the notebook you referred to:

![image](https://user-images.githubusercontent.com/22635213/126934824-c8b81c7e-59f9-4532-84d5-5ab090c859dc.png)

In the text: _It is better to consider a renormalized gradient, which corresponds to using the inner product associated to the measure a on the deformation field_. This "renormalized" gradient is basically the set of points cancelling the quadratic energy that results from the Danskin approx of the reg OT cost. It's not the "true" gradient.

I have modified your NB and you can see both gradients agree in the sense that they are a constant factor of each other: [link](https://colab.research.google.com/drive/1pFWbZeFSOskFTY2qMdRu8mStu9TRw7eE?usp=sharing)

you can also check at the end that the gradient returned by OTT is the one that corresponds strictly speaking to that of the energy.

I just wanted to thank you again for doing these checks on your side, this is really helpful for us.

**Comment by ersisimou** [comment]
Hi @marcocuturi ,

Thank you for the response!

First, let me start by saying that I provided the links to the jupyter notebook and the test, just to ensure that I have given the credits, since I just copy-pasted code from these resources in the collab I shared. :)

That being said, and since you mentioned the text from the notebook, I guess my question is: "Is it better to consider a renormalized gradient or necessary"? I understand that for a computer, the first gradient in your screenshot above will be a matrix of the same shape as the matrix with the current positions z and that one might just implement the gradient step to update the positions z using that unormalized gradient without getting an error. However, the way I see it, if one looks at this as a physical problem (where for instance a is really mass), using the the unnormalized gradient would be like updating the positions using a vector of mass*position. This normalization is actually also present in the computation of the [positions](http://proceedings.mlr.press/v32/cuturi14.pdf) of the barycenter at a Lagrangian discretization setting using gradient descent and a squared Euclidean distance cost. Therefore, it seems to me that this normalization with the distribution of mass (here the vector a) is actually the natural thing to do in order to update the positions, right? Also, if the distribution of mass is not uniform, I understand that this normalization will be more than just a scaling. 

Bringing this back to the context of this toolbox however, how should one update the positions of mass using the gradient returned by jax.grad on the Sinkhorn output? If the values returned by jax.grad correspond to the unnormalized gradient, then should we normalize it with the weights a? If yes, then I notice that the normalized gradient is actually 2 times the gradient returned by the analytic computation (aka computation in the jupyter notebook in my previous post or paper in this post). 

Thanks again for your kind and prompt interaction on the Github issues of this toolbox. It is a pleasure to use it :)

**Comment by marcocuturi** [comment]
To your first question:
What is called the "normalized" gradient is indeed the update in Eq. (8) in [my 2014 paper](http://proceedings.mlr.press/v32/cuturi14.pdf), which is also known as a barycentric projection. This corresponds to the solution of the quadratic upper bound on the transport cost you can get by freezing $P^*$. That being said, I think this is not strictly speaking a gradient update, but rather using a closed form solution for a quadratic problem (as in a block coordinate scheme). Although Gabriel called it a "normalized" gradient, to me it's slightly misleading. For instance, if
$f(x) := (x-a)^2$, the gradient value is $f'(x) = 2(x-a)$ and the "normalized gradient" would be $a$ itself (the value cancelling the gradient of the quadratic form).

In terms of choice, I prefer to follow the more general convention that $ \nabla f(x)$ must be such that $f(x+h) - f(x) \approx \langle h, \nabla f(x)$, $f$ being the regularized OT cost.

For the second question:
I guess it's an open debate. Another choice could for instance be to meet half-way, $ x \leftarrow (x + y)/2$. Since this is a non convex problem it's a tough question :) It's also related to what exactly you are minimizing. If it's a barycenter, then maybe using the average of those $y$ values makes sense indeed (mix of "normalized gradients"), but if it's anything more complicated maybe sticking with another first order method (stochastic, Adagrad, ADAM etc..) could make sense.

**Comment by ersisimou** [comment]
Hi @marcocuturi ,

Thank you for the answers! I will close this issue.
